library/sha256.c

```diff
@@ -35,6 +35,28 @@
 #include "mbedtls/platform.h"
 
 #if defined(__aarch64__)
+
+/*
+ * Clang is disabled due to https://reviews.llvm.org/D131064
```

Comment:
Sorry, I don't understand, after reading that link a couple of times. Can you explain why that PR means we can no longer use clang to compile accelerated SHA-2? (I spent a lot of time testing all combinations of SHAxxx_USE_A64_CRYPTO_* on multiple versions of clang, and it did used to work - this seems to be a regression

Reply:
> Can you explain why that PR means we can no longer use clang to compile accelerated SHA-2? 

It still works. I only disable it when `__ARM_FEATURE_CRYPTO` is disabled. With `-march=armv8-a+crypto`, clang still work.  
Sorry, I forget add the version checks when `__ARM_FEATURE_CRYPTO` is enabled.

Reply:
> Sorry, I don't understand, after reading that link a couple of times. 

[clang](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:22,endLineNumber:8,positionColumn:22,positionLineNumber:8,selectionStartColumn:22,selectionStartLineNumber:8,startColumn:22,startLineNumber:8),source:'%0A%23if+defined(__clang__)%0A%23pragma+clang+attribute+push+(__attribute__((target(%22crypto%22))),+apply_to%3Dfunction)%0A%23elif+defined(__GNUC__)%0A%23pragma+GCC+target+(%22arch%3Darmv8-a%2Bcrypto%22)%0A%23endif%0A%0A%23include+%3Carm_neon.h%3E%0Auint8x16_t+aesce_encrypt_block(uint8x16_t+block,%0A++++++++++++++++++++++++++++++++++++++unsigned+char+*keys,%0A++++++++++++++++++++++++++++++++++++++int+rounds)%0A%7B%0A++++for+(int+i+%3D+0%3B+i+%3C+rounds+-+1%3B+i%2B%2B)+%7B%0A++++++++block+%3D+vaeseq_u8(block,+vld1q_u8(keys+%2B+i+*+16))%3B%0A++++++++/*+AES+mix+columns+*/%0A++++++++block+%3D+vaesmcq_u8(block)%3B%0A++++%7D%0A%0A++++/*+AES+single+round+encryption+*/%0A++++block+%3D+vaeseq_u8(block,+vld1q_u8(keys+%2B+(rounds+-1)+*+16))%3B%0A%0A++++/*+Final+Add+(bitwise+Xor)+*/%0A++++block+%3D+veorq_u8(block,+vld1q_u8(keys+%2B+rounds++*+16))%3B%0A%0A++++return+block%3B%0A%7D%0A%23if+defined(__clang__)%0A%23pragma+clang+attribute+pop%0A%23endif%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:42.64782836211409,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armv8-cclang1500,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+armv8-a+clang+15.0.0+(Editor+%231)',t:'0')),k:24.018838304552595,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'armv8-a+clang+15.0.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+armv8-a+clang+15.0.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) report function undefined error.  And [gcc](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:6,endLineNumber:17,positionColumn:6,positionLineNumber:17,selectionStartColumn:6,selectionStartLineNumber:17,startColumn:6,startLineNumber:17),source:'%0A%23if+defined(__clang__)%0A%23pragma+clang+attribute+push+(__attribute__((target(%22crypto%22))),+apply_to%3Dfunction)%0A%23elif+defined(__GNUC__)%0A%23pragma+GCC+target+(%22arch%3Darmv8-a%2Bcrypto%22)%0A%23endif%0A%0A%23include+%3Carm_neon.h%3E%0Auint8x16_t+aesce_encrypt_block(uint8x16_t+block,%0A++++++++++++++++++++++++++++++++++++++unsigned+char+*keys,%0A++++++++++++++++++++++++++++++++++++++int+rounds)%0A%7B%0A++++for+(int+i+%3D+0%3B+i+%3C+rounds+-+1%3B+i%2B%2B)+%7B%0A++++++++block+%3D+vaeseq_u8(block,+vld1q_u8(keys+%2B+i+*+16))%3B%0A++++++++/*+AES+mix+columns+*/%0A++++++++block+%3D+vaesmcq_u8(block)%3B%0A++++%7D%0A%0A++++/*+AES+single+round+encryption+*/%0A++++block+%3D+vaeseq_u8(block,+vld1q_u8(keys+%2B+(rounds+-1)+*+16))%3B%0A%0A++++/*+Final+Add+(bitwise+Xor)+*/%0A++++block+%3D+veorq_u8(block,+vld1q_u8(keys+%2B+rounds++*+16))%3B%0A%0A++++return+block%3B%0A%7D%0A%23if+defined(__clang__)%0A%23pragma+clang+attribute+pop%0A%23endif%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:42.64782836211409,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:carm64g1220,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2+',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+ARM64+gcc+12.2+(Editor+%231)',t:'0')),k:24.018838304552595,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'armv8-a+clang+15.0.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+ARM64+gcc+12.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) can pass it. 

That is an bug of `arm_neon.h` from clang. https://reviews.llvm.org/D131064 is to change `arm_neon.h`.   If you do not add command line flags , some intrinsic functions are not available. That's the bug.

Reply:
It is reverted.